### DISCUSSION TITLE
Deploying Nest JS website on Vercel

### DISCUSSION THREAD

**Discussion by st3fus** [vercel/community#2323] [Help] [answered]
### Summary

So I've been trying to deploy a website on Vercel that's done in Nest JS, the html is served from within views folder. Build passes with no errors, but on production link, I either get a Serverless Function failed error 500 (it cant find a module for firebase.service but the module is installed and the imports are correct) or when I remove the build arguments from vercel.json I get a 404. Anyone had experience with this before?

### Additional information

```
{
    "version": 2,
	"builds": [
        {
            "src": "src/main.ts",
            "use": "@vercel/node"
        }
    ],
    "routes": [
        {
            "src": "/(.*)",
            "dest": "src/main.ts",
            "methods": [
                "GET",
                "POST",
                "PUT",
                "DELETE"
            ]
        }
    ]
}
```

package.json
```
{
  "name": "taskadu",
  "version": "0.0.1",
  "description": "",
  "author": "",
  "private": true,
  "license": "UNLICENSED",
  "scripts": {
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:dev": "nest start --watch",
    "start:debug": "nest start --debug --watch",
    "start:prod": "node dist/main",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
  "dependencies": {
    "@nestjs-modules/mailer": "^1.8.1",
    "@nestjs/common": "^9.0.0",
    "@nestjs/config": "^2.3.1",
    "@nestjs/core": "^9.0.0",
    "@nestjs/mapped-types": "*",
    "@nestjs/passport": "^9.0.1",
    "@nestjs/platform-express": "^9.2.1",
    "class-transformer": "^0.5.1",
    "class-validator": "^0.14.0",
    "firebase": "^9.17.1",
    "firebase-admin": "^11.5.0",
    "hbs": "^4.2.0",
    "hbs-utils": "0.0.4",
    "jquery": "^3.6.3",
    "nestjs-form-data": "^1.8.7",
    "passport": "^0.6.0",
    "passport-firebase-jwt": "^1.2.1",
    "reflect-metadata": "^0.1.13",
    "rxjs": "^7.2.0",
    "uuid": "^9.0.0"
  },
  "devDependencies": {
    "@nestjs/cli": "^9.0.0",
    "@nestjs/schematics": "^9.0.0",
    "@nestjs/testing": "^9.0.0",
    "@types/express": "^4.17.13",
    "@types/jest": "29.2.4",
    "@types/node": "18.11.18",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@typescript-eslint/parser": "^5.0.0",
    "eslint": "^8.0.1",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "jest": "29.3.1",
    "prettier": "^2.3.2",
    "run-script-webpack-plugin": "^0.1.1",
    "source-map-support": "^0.5.20",
    "supertest": "^6.1.3",
    "ts-jest": "29.0.3",
    "ts-loader": "^9.2.3",
    "ts-node": "^10.0.0",
    "tsconfig-paths": "4.1.1",
    "typescript": "^4.7.4",
    "webpack": "^5.75.0",
    "webpack-node-externals": "^3.0.0"
  },
  "jest": {
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "rootDir": "src",
    "testRegex": ".*\\.spec\\.ts$",
    "transform": {
      "^.+\\.(t|j)s$": "ts-jest"
    },
    "collectCoverageFrom": [
      "**/*.(t|j)s"
    ],
    "coverageDirectory": "../coverage",
    "testEnvironment": "node"
  }
}
```
The error I'm getting:
```
Cannot find module 'src/shared/services/firebase.service'
Require stack:
- /var/task/src/profile/profile.service.js
- /var/task/src/profile/profile.module.js
- /var/task/src/app.module.js
- /var/task/src/main.js
Did you forget to add it to "dependencies" in `package.json`?
RequestId: adf1bfac-8549-4b73-866e-d10efdec4e16 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

But NPM install and build were successful.
### Example

https://taskadu.vercel.app/

**Comment by st3fus** [comment]
Anyone?

**Reply by Reghima** [reply]
i havethe same issue too

**Reply by h7ml** [reply]
i havethe same issue too

**Reply by ngquocvu** [reply]
Me too

**Reply by yasilvalmeida** [reply]
I'm facing the same issue, I try to add aliases in tsconfig.json but not working

**Reply by mohrazzak** [reply]
you must use relative imports

**Reply by thegdsks** [reply]
Smae

**Comment by Gruce** [comment]
Same here.

**Comment by amyegan** [comment]
Does anyone here have a public repo that I can run to get more details about the error?

**Reply by Gruce** [reply]
You can just install fresh nestjs.

**Reply by rafamsl** [reply]
These are my repo. 

I am getting this error on HBS engine

https://github.com/rafamsl/language-guesser3

[31m[Nest] 8  - [39m07/02/2023, 8:37:48 PM [31m  ERROR[39m [38;5;3m[ExceptionsHandler] [39m[31mCannot find module 'hbs'

I also had a second repo which was getting this error on a simple class import

https://github.com/rafamsl/language-guesser2

Cannot find module 'src/dto/create-dailySampleDTO.dto'

**Reply by Gruce** [reply]
> These are my repo. 
> 
> I am getting this error on HBS engine
> 
> https://github.com/rafamsl/language-guesser3
> 
> [31m[Nest] 8  - [39m07/02/2023, 8:37:48 PM [31m  ERROR[39m [38;5;3m[ExceptionsHandler] [39m[31mCannot find module 'hbs'
> 
> I also had a second repo which was getting this error on a simple class import
> 
> https://github.com/rafamsl/language-guesser2
> 
> Cannot find module 'src/dto/create-dailySampleDTO.dto'

Please use relative path instead of relative.

**Reply by edgexie** [reply]
I have the same error,
but I change to use relative path, It's ok. why? any ideas?

**Comment by rafamsl** [comment]
came here for answers, left more intrigued. same issue as well.

**Comment by tofiqquadri** [comment, marked as answer]
I found the solution to this error. If you have this exact error:

```
Cannot find module 'src/ANY_PATH'
Require stack:
```

1. Then go to your code and wherever you have made an import with 'src/YOUR_FILE_PATH' change that to relative import like: '../../YOUR_FILE_PATH'

2. In your .gitignore file remove /dist and delete the dist folder.

3. Use vercel.json exactly as used by @st3fus 

Commit the code and deploy it on vercel. It should work.

**Reply by chhornponleu** [reply]
I faced the same issue and this is the solution. I removed `"baseUrl": "./",` from `tsconfig.json` and then fix the import error. Then every just works fine.

**Comment by behrangsa** [comment]
A bit off topic, but I am a bit confused.

Isn't Vercel a platform for deploying frontends + simple/BFF APIs?

**Comment by Akshithpottigari** [comment]
I have removed the baseUrl property from tsconfig.json, and then adjusting the import statements in all the files, and it worked

**Reply by aziyatali** [reply]
Thank buddy, this was helpful

**Reply by Akshithpottigari** [reply]
You're welcome 👍

**Comment by sahariar-safin** [comment]
If you follow the steps outlined in the article, you will be able to successfully deploy NestJS on Vercel 🚀

#https://www.todayunboxed.com/deploy-your-nestjs-app-like-a-pro-with-vercel-in-2024-step-by-step-guide #
